### ISSUE TITLE
Supporting extensions?

### ISSUE THREAD

**Issue by alexsnaps** [cel-rust/cel-rust#103]
Wonder how, if at all, this library would support [extensions](https://github.com/google/cel-go/tree/master/ext) in ways the golang implementation does? We could have the need for the [`Strings()`](https://github.com/google/cel-go/blob/master/ext/strings.go#L289), how would that best be supported here, if that's interesting?

**Comment by clarkmcc** [comment]
It looks like these are just CEL functions that are namespaced and not available in the interpreter by default. The biggest missing piece in the library is the ability to namespace functions. You can already write your own functions and load them into the interpreter so I'm guessing your questions or more about (1) namespacing, and (2) whether we can provide these extensions out of the box.

I don't have any reservations about supporting this, though I do have several other featuers and PR reviews that are spec-compliance-related that I think are higher priority than this. That being said, if you wanted to take a stab at implementing, I'd be more than happy to review.

**Comment by alexsnaps** [comment]
Yeah :) I was thinking of providing this straight to this crate... Something along side how its done in Go

```go
	env, err := cel.NewEnv(Strings())
```

Tho, unsure about the namespacing? [`Strings`](https://github.com/google/cel-go/blob/master/ext/strings_test.go#L43) are straight methods on strings... are they not?

**Comment by alexsnaps** [comment]
Not so [all of them](https://github.com/google/cel-go/blob/master/ext/protos_test.go#L38-L50), got it!

**Comment by alexsnaps** [comment]
I haven't dropped the ball on this... I ended up [implementing this straight on our usage of this](https://github.com/Kuadrant/wasm-shim/pull/132/files) only because I didn't see how to best make this available from an API perspective in a way that's seamless enough quite yet. But in case it can be of any help to anyone, `cel::strings` on that PR has all _initial_ string extensions.

**Comment by howardjohn** [comment]
What about something like:

```rust
let ctx = Context::default().with_regex().with_time().with_string()
```

Or

```rust
let ctx = Context::default().add_function_set(functions::strings())
```

Where `add_function_set` accepts a trait that just has `fn add_functions(&mut Context)`.

Then we can just expose the extension sets, maybe under a config flags?

**Comment by alexsnaps** [comment]
@howardjohn I'm currently working on implementing the CEL stdlib with proper overloads. Expect a PR in the couple of days. Once that is done, we'll be able to support these extension (and be open to custom libs) much easier.

**Comment by clarkmcc** [comment]
@alexsnaps not sure what you have in mind there, but I will say (and maybe it's just my lack of understanding) that the way to configure methods and functions with cel-go is pretty tedious compared to how it works in this crate today. Do you intend to preserve sort of the Rust-type function experience, e.g.:
```rust
ctx.add_function("my_func", |a: i64, b: i64| a + b)
```

**Comment by alexsnaps** [comment]
That's more of the public API, which I wasn't really talking about here. There I'd be in favor of whatever works best, for some definition of "best". I was actually more thinking of the internals here and how `FuncCall` resolution works.

**Comment by doxxx93** [comment]
think this might be unblocked now, the stdlib rework you were waiting on landed in #283.

Is `Context::with_library(...)` from #253 still the plan now that `Env` is in, or would libraries register on `Env` instead?

**Comment by alexsnaps** [comment]
It would ultimately need to be on Env. we can't expect to have a context (ie variables) until evaluation time. But we want to be able to check the AST for correctness. In this case, make sure all calls do have matching overloads. Does that make sense?

**Comment by alexsnaps** [comment]
Started [this discussion](https://github.com/cel-rust/cel-rust/discussions/323) on that wider subject

**Comment by doxxx93** [comment]
Yeah, makes sense. Registration has to be resolvable before evaluation if the check step is going to verify overloads.

Taking it to #323.

**Comment by alexsnaps** [comment]
Once we have #276 & #335 done, with #353 & #334 - this issue would "just" be about tying it all together in a nice API for users to call, something like:

```rust
let mut env = Env::default();
env.add_extension(some::crate::awesome_cel_extension);
```
